R/dplyr_methods.R

```diff
@@ -911,40 +911,144 @@ slice_sample.Seurat <- function(.data, ..., n = NULL, prop = NULL, by = NULL, we
   # Solve CRAN NOTES
   cell = NULL
   . = NULL
-  
+
   lifecycle::signal_superseded("1.0.0", "sample_n()", "slice_sample()")
-  
+
   if(!is.null(n))
     new_meta =
-      .data[[]] %>%  
-      as_tibble(rownames = c_(.data)$name) %>% 
+      .data[[]] %>%
+      as_tibble(rownames = c_(.data)$name) %>%
       dplyr::slice_sample(..., n = n, by = by, weight_by = weight_by, replace = replace)
   else if(!is.null(prop))
     new_meta =
-    .data[[]] %>%  
-    as_tibble(rownames = c_(.data)$name) %>% 
+    .data[[]] %>%
+    as_tibble(rownames = c_(.data)$name) %>%
     dplyr::slice_sample(..., prop=prop, by = by, weight_by = weight_by, replace = replace)
   else
     stop("tidyseurat says: you should provide `n` or `prop` arguments")
-  
+
   count_cells = new_meta %>% select(!!c_(.data)$symbol) %>% count(!!c_(.data)$symbol)
-  
+
   # If repeted cells
   if(count_cells$n %>% max() %>% gt(1)){
     message("tidyseurat says: When sampling with replacement a data frame is returned for independent data analysis.")
-    .data %>% 
-      as_tibble() %>% 
+    .data %>%
+      as_tibble() %>%
       right_join(new_meta %>% select(!!c_(.data)$symbol),  by = c_(.data)$name)
   }  else{
     new_obj = subset(.data,   cells = new_meta %>% pull(!!c_(.data)$symbol))
-    new_obj@meta.data = 
-      new_meta %>% 
+    new_obj@meta.data =
+      new_meta %>%
       data.frame(row.names=pull(.,!!c_(.data)$symbol), check.names = FALSE) %>%
-      select(- !!c_(.data)$symbol) 
+      select(- !!c_(.data)$symbol)
     new_obj
   }
-  
-  
+
+
+}
+
+#' @rdname dplyr-methods
+#' @name slice_head
+#' @importFrom dplyr slice_head
+#' @examples
+#'
+#' # First rows based on existing order
+#' pbmc_small |> slice_head(n = 5)
+NULL
+
+#' @export
+slice_head.Seurat <- function(.data, ..., n, prop, by = NULL) {
+    new_meta <- dplyr::slice_head(.data[[]], ..., n = n, prop = prop, by = {{by}})
```

Comment:
To save memory for huge dataset, could we 
- somehow select 0 columns from this dataset .data[[]] ? Would the dataset still be valid?
- instead of new_meta we could just get rownames ?

What if the metadata does not have row names? For other tidy adapters we have used the trick of rowid_to_column(name = "row_number___"), then used this index to subset Seurat object .data[...,]

This would apply to all new functions.

Reply:
What do you think of this,

a. https://github.com/stemangiola/tidyseurat/pull/68/commits/1d2ed45aa10733367da2d6733ebfe0d709557630 extract entire metadata first but then subset to needed columns, for `slice` the only column needed would be the optional argument `{{ by }}`. Other slice functions like _min _max could have more variables

and use `rowid_to_column` in case there are no rownames

another approach:

b. https://github.com/stemangiola/tidyseurat/pull/68/commits/d0b85918f01f1d0411dfb21878118ca80b95f747 construct an empty data.frame with no columns using rownames of .data

add in necessary variables for slice, so `{{ by }}` and arrange_by variables

perform slice operation and return row index / row names


b. is more memory efficient than a., due to never extracting the entire metadata df. But the implementation is a bit more verbose and complex

Reply:
I would say `a` is good 

- select columns
- `rowid_to_column` in any case
- `slice_*`

Great!